src/software/ai/hl/stp/tactic/shadow_enemy/shadow_enemy_fsm.cpp

```diff
@@ -38,15 +39,41 @@ Point ShadowEnemyFSM::findBlockShotPoint(const Robot &robot, const Field &field,
 bool ShadowEnemyFSM::enemyThreatHasBall(const Update &event)
 {
     std::optional<EnemyThreat> enemy_threat_opt = event.control_params.enemy_threat;
+
     if (enemy_threat_opt.has_value())
     {
-        return enemy_threat_opt.value().has_ball;
-    };
-    LOG(WARNING) << "Enemy threat not initialized for robot " << event.common.robot.id()
-                 << "\n";
+        bool near_ball = distance(event.common.world_ptr->ball().position(),
+                                  enemy_threat_opt.value().robot.position()) < 0.22;
+
+        return near_ball;
+    }
     return false;
 }
 
+bool ShadowEnemyFSM::blockedShot(const Update &event)
+{
+    auto ball_position = event.common.world_ptr->ball().position();
+    Ray shot_block_direction(ball_position,
+                             event.common.robot.position() - ball_position);
+    Segment goalLine(event.common.world_ptr->field().friendlyGoal().posXNegYCorner(),
+                     event.common.world_ptr->field().friendlyGoal().posXPosYCorner());
+
```

Comment:
```suggestion
    Segment goalLine(event.common.world_ptr->field().friendlyGoal().negXNegYCorner(),
                     event.common.world_ptr->field().friendlyGoal().negXPosYCorner());
 
 `posXPosYCorner` will give you the top of the goal box, and not the back goal line, like you intend
```

we also have `friendlyGoalpostNeg` and `friendlyGoalpostPos` helpers instead of computing it manually here